source-code/core/app/src/createInlang.ts

```diff
@@ -0,0 +1,220 @@
+			lintMessages({
+				config: configValue,
+				messages: msgs,
+				query,
+				rules: resolvedModules()!.data.lintRules.data,
+			}).then((report) => {
+				setLintReports(report.data)
+				setLintErrors(report.errors)
+
+				markLintReportsAsInitialized()
+			})
+		})
+
+		// -- app ---------------------------------------------------------------
+
+		await initialized.catch((e) => {
+			throw e
+		})
+
+		return {
+			meta: {
+				plugins: createSubscribable(() => resolvedModules()!.data.meta.plugins),
+				lintRules: createSubscribable(() => resolvedModules()!.data.meta.lintRules),
+			},
+			errors: createSubscribable(() => [
+				...resolvedModules()!.errors,
+				...resolvedModules()!.data.plugins.errors,
+				...resolvedModules()!.data.lintRules.errors,
+				...(lintErrors() || []),
+			]),
+			config: createSubscribable(() => config()!),
+			setConfig: setConfig,
+			lint: {
+				init: initLint,
+				reports: createSubscribable(() => {
+					const reports = lintReports()
+					// TODO: improve error
+					if (!reports) return []
+					return reports
+				}),
+			},
+			appSpecificApi: createSubscribable(
+				() => resolvedModules()!.data.plugins.data!.appSpecificApi,
+			),
+			query: {
+				messages: query,
+			},
+		} satisfies InlangInstance
+	})
+}
+
+const loadConfig = async (args: { configPath: string; nodeishFs: NodeishFilesystemSubset }) => {
+	const { data: configFile, error: configFileError } = await tryCatch(
+		async () => await args.nodeishFs.readFile(args.configPath, { encoding: "utf-8" }),
+	)
+	if (configFileError)
+		throw new ConfigPathNotFoundError(`Could not locate config file in (${args.configPath}).`, {
+			cause: configFileError,
+		})
+
+	const { data: parsedConfig, error: parseConfigError } = tryCatch(() => JSON.parse(configFile!))
+	if (parseConfigError)
+		throw new ConfigSyntaxError(`The config is not a valid JSON file.`, {
+			cause: parseConfigError,
+		})
+
+	const typeErrors = [...ConfigCompiler.Errors(parsedConfig)]
+	if (typeErrors.length > 0) {
+		throw new InvalidConfigError(`The config is invalid according to the schema.`, {
+			cause: typeErrors,
+		})
+	}
+
+	return Value.Cast(InlangConfig, parsedConfig)
+}
+
+const loadModules = async (args: {
+	config: InlangConfig
+	nodeishFs: NodeishFilesystemSubset
+	_import?: ImportFunction
+}) =>
+	resolveModules({
+		config: args.config,
+		nodeishFs: args.nodeishFs,
+		_import: args._import,
+	})
+
+const createAwaitable = () => {
+	let resolve: () => void
+	let reject: () => void
+
+	const promise = new Promise<void>((res, rej) => {
+		resolve = res
+		reject = rej
+	})
+
+	return [promise, resolve!, reject!] as [
+		awaitable: Promise<void>,
+		resolve: () => void,
+		reject: (e: unknown) => void,
+	]
+}
+
+// TODO: create global util type
+type MaybePromise<T> = T | Promise<T>
+
+const makeTrulyAsync = <T>(fn: MaybePromise<T>): Promise<T> => (async () => fn)()
+
+
+// TODO: how do we unsubscribe? Do we need that?
```

Comment:
as i understand, no. we don't need to unsubscribe because the reactivity is handled within the inlang instance and is thereby auto tracked via scopes.

Reply:
@ivanhofer i guess we will see if we run into memory leaks :D